# Проектная работа по модулю 'Python'.

Датасет 'Employee Attrition' взят с  https://www.kaggle.com/datasets/patelprashant/employee-attrition/data.

### Описание данных.

Этот набор данных содержит 34 признака, которые могут оказывать влияние на выбытие сотрудника из штата (Attrition Выбыл из штата). 
* Age возраст
* BusinessTravel Командировки
* DailyRate Дневная норма
* Department Отделение
* DistanceFromHome Расстояние от дома
* Education Образование (1 - «Ниже колледжа», 2 - «Колледж», 3 - «Бакалавр», 4 - «Магистр», 5 - «Доктор»)
* EducationField Область образования
* EmployeeCount Количество сотрудников
* EmployeeNumber Номер сотрудника
* EnvironmentSatisfaction Удовлетворенность окружающей средой (1 - «Низкая», 2 - «Средняя», 3 - «Высокая», 4 - «Очень высокая»)
* Gender Пол
* HourlyRate Почасовая ставка
* JobInvolvement Вовлеченность в работу (1 - «Низкий», 2 - «Средний», 3 - «Высокий», 4 - «Очень высокий»)
* JobLevel Уровень работы
* JobRole Должность
* JobSatisfaction Удовлетворение от работы (1 - «Низкая», 2 - «Средняя», 3 - «Высокая», 4 - «Очень высокая»)
* MaritalStatus Семейное положение
* MonthlyIncome Ежемесячный доход
* MonthlyRate Ежемесячная ставка
* NumCompaniesWorked Количество компаний, в которых работал
* Over18 Старше 18
* OverTime Сверхурочные работы
* PercentSalaryHike Процент повышения зарплаты
* PerformanceRating рейтинг производительности (1 - «Низкий», 2 - «Хороший», 3 - «Отличный», 4 - «Выдающийся»)
* RelationshipSatisfaction Удовлетворенность отношениями (1 - «Низкая», 2 - «Средняя», 3 - «Высокая», 4 - «Очень высокая»)
* StandardHours Стандартные часы
* StockOptionLevel Уровень опционов на акции
* TotalWorkingYears Всего рабочих лет
* TrainingTimesLastYear Время тренировок в прошлом году
* WorkLifeBalance Баланс между работой и личной жизнью (1 - «Плохо», 2 - «Хорошо», 3 - «Лучше», 4 - «Наилучшее»)
* YearsAtCompany Лет в компании
* YearsInCurrentRole Лет в нынешней роли
* YearsSinceLastPromotion Лет с момента последнего повышения
* YearsWithCurrManager Лет с нынешним менеджером

Дополнительное описание, которое поможет понять особенности:

Почасовая ставка: почасовому работнику выплачивается заработная плата за каждый отработанный час, в отличие от наемного работника.

EmployeeNumber: Номер сотрудника — это уникальный номер, который был присвоен каждому действующему и бывшему государственному служащему и выборному должностному лицу в базе данных должностей и персонала (PPDB).

JobLevel: уровни должностей, также известные как разряды и классификации должностей, устанавливают уровень ответственности и ожидания от ролей в вашей организации. Они могут дополнительно определяться влиянием, стажем, знаниями, навыками или должностью и часто связаны с диапазоном заработной платы. 

StockOptionLevel: опционы на акции для сотрудников, также известные как ESO, представляют собой опционы на акции компании, предоставляемые работодателем определенным сотрудникам. Обычно они предоставляются лицам, занимающим руководящие или офицерские должности. Опционы на акции дают сотруднику право купить определенное количество акций по определенной цене в течение определенного периода времени. Опционы обычно также имеют дату истечения срока действия, до которой опционы должны быть исполнены, иначе они станут бесполезными.


### Вопросы.

1.	Как на сокращение штата сотрудников влияет пол?
2.	Какой средний возраст сотрудников, выбывших из штата?

Определим исследуемую группу.

3.	Как влияет наличие частых командировок на выбытие из штата сотрудников исследуемой группы? 
4.	Узнай доли выбывших сотрудников исследуемой группы по каждому отделу
5.	Узнай наиболее часто встречаемую должность среди тех, кто выбыл из штата в исследуемой группе. 
6.	Сотрудники с каким семейным положением в исследуемой группе выбывают из штата чаще?
7.	Какой средний ежемесячный доход в исследуемой группе? Узнай долю сотрудников, имеющих сверхурочные работы, среди тех, кто выбыл из штата в исследуемой группе и чей ежемесячный доход ниже найденного среднего.
8.	Сколько сотрудников исследуемой группы служат в текущей компании меньше 10 лет?  Какая доля таких сотрудников выбыла из штата?
9.	Правда ли, что среди выбывших сотрудников исследуемой группы больше всего тех, чье образование ниже магистра?
10.	 Среди выбывших сотрудников исследуемой группы с какой областью образования больше? 
11.	 Правда ли, что среди выбывших сотрудников исследуемой группы больше всего тех, у кого низкая удовлетворенность работой? 

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('datasets/WA_Fn-UseC_-HR-Employee-Attrition.csv')
df.sample(5)

Age Attrition     BusinessTravel  DailyRate              Department  \
578    38        No  Travel_Frequently        240  Research & Development   
585    23       Yes      Travel_Rarely       1243  Research & Development   
1174   25        No  Travel_Frequently        772  Research & Development   
823    29        No  Travel_Frequently        490  Research & Development   
564    45        No      Travel_Rarely        954                   Sales   

      DistanceFromHome  Education    EducationField  EmployeeCount  \
578                  2          4     Life Sciences              1   
585                  6          3     Life Sciences              1   
1174                 2          1     Life Sciences              1   
823                 10          3     Life Sciences              1   
564                  2          2  Technical Degree              1   

      EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
578              803  ...                         4            80   
585              811  ...                         3            80   
1174            1653  ...                         3            80   
823             1143  ...                         4            80   
564              783  ...                         1            80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
578                  0                 17                      2   
585                  2                  1                      2   
1174                 2                  7                      6   
823                  2                  8                      2   
564                  0                  9                      3   

     WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
578                3              15                  7   
585                3               0                  0   
1174               3               7                  7   
823                2               7                  5   
564                3               8                  7   

      YearsSinceLastPromotion  YearsWithCurrManager  
578                         4                    12  
585                         0                     0  
1174                        0                     7  
823                         1                     1  
564                         3                     1  

[5 rows x 35 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [4]:
df.isnull().any().any()

False

В данных нет пропусков

In [5]:
df.describe()

Age    DailyRate  DistanceFromHome    Education  EmployeeCount  \
count  1470.000000  1470.000000       1470.000000  1470.000000         1470.0   
mean     36.923810   802.485714          9.192517     2.912925            1.0   
std       9.135373   403.509100          8.106864     1.024165            0.0   
min      18.000000   102.000000          1.000000     1.000000            1.0   
25%      30.000000   465.000000          2.000000     2.000000            1.0   
50%      36.000000   802.000000          7.000000     3.000000            1.0   
75%      43.000000  1157.000000         14.000000     4.000000            1.0   
max      60.000000  1499.000000         29.000000     5.000000            1.0   

       EmployeeNumber  EnvironmentSatisfaction   HourlyRate  JobInvolvement  \
count     1470.000000              1470.000000  1470.000000     1470.000000   
mean      1024.865306                 2.721769    65.891156        2.729932   
std        602.024335                 1.093082    20.329428        0.711561   
min          1.000000                 1.000000    30.000000        1.000000   
25%        491.250000                 2.000000    48.000000        2.000000   
50%       1020.500000                 3.000000    66.000000        3.000000   
75%       1555.750000                 4.000000    83.750000        3.000000   
max       2068.000000                 4.000000   100.000000        4.000000   

          JobLevel  ...  RelationshipSatisfaction  StandardHours  \
count  1470.000000  ...               1470.000000         1470.0   
mean      2.063946  ...                  2.712245           80.0   
std       1.106940  ...                  1.081209            0.0   
min       1.000000  ...                  1.000000           80.0   
25%       1.000000  ...                  2.000000           80.0   
50%       2.000000  ...                  3.000000           80.0   
75%       3.000000  ...                  4.000000           80.0   
max       5.000000  ...                  4.000000           80.0   

       StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
count       1470.000000        1470.000000            1470.000000   
mean           0.793878          11.279592               2.799320   
std            0.852077           7.780782               1.289271   
min            0.000000           0.000000               0.000000   
25%            0.000000           6.000000               2.000000   
50%            1.000000          10.000000               3.000000   
75%            1.000000          15.000000               3.000000   
max            3.000000          40.000000               6.000000   

       WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
count      1470.000000     1470.000000         1470.000000   
mean          2.761224        7.008163            4.229252   
std           0.706476        6.126525            3.623137   
min           1.000000        0.000000            0.000000   
25%           2.000000        3.000000            2.000000   
50%           3.000000        5.000000            3.000000   
75%           3.000000        9.000000            7.000000   
max           4.000000       40.000000           18.000000   

       YearsSinceLastPromotion  YearsWithCurrManager  
count              1470.000000           1470.000000  
mean                  2.187755              4.123129  
std                   3.222430              3.568136  
min                   0.000000              0.000000  
25%                   0.000000              2.000000  
50%                   1.000000              3.000000  
75%                   3.000000              7.000000  
max                  15.000000             17.000000  

[8 rows x 26 columns]

In [6]:
df.describe(include="object")

Attrition BusinessTravel              Department EducationField Gender  \
count       1470           1470                    1470           1470   1470   
unique         2              3                       3              6      2   
top           No  Travel_Rarely  Research & Development  Life Sciences   Male   
freq        1233           1043                     961            606    882   

                JobRole MaritalStatus Over18 OverTime  
count              1470          1470   1470     1470  
unique                9             3      1        2  
top     Sales Executive       Married      Y       No  
freq                326           673   1470     1054

#### 1.	Как на сокращение штата сотрудников влияет пол?

In [7]:
df_gender = pd.DataFrame(df.groupby(['Gender'])[['Attrition']].value_counts().reset_index())
df_gender_attr = df_gender[df_gender['Attrition'] == 'Yes']
df_gender_attr

Gender Attrition  count
1  Female       Yes     87
3    Male       Yes    150

In [8]:
female_attr = df_gender_attr[df_gender_attr['Gender'] == 'Female']['count'].values[0] / df_gender_attr['count'].sum()
female_attr.round(2)

0.37

In [9]:
male_attr = df_gender_attr[df_gender_attr['Gender'] == 'Male']['count'].values[0] / df_gender_attr['count'].sum()
male_attr.round(2)

0.63

ответ: больше всего выбывших среди мужчин, почти в 2 раза больше, чем женщин

#### 2.	Какой средний возраст сотрудников, выбывших из штата?

In [10]:
df['Age'].mean().round()

37.0

In [11]:
df['Age'].describe()

count    1470.000000
mean       36.923810
std         9.135373
min        18.000000
25%        30.000000
50%        36.000000
75%        43.000000
max        60.000000
Name: Age, dtype: float64

ответ: средний возраст сотрудников, выбывших из штата, равен 37 лет

## Определим, что мы исследуем группу - мужчины возраста 30-43 года.

#### 3.	Как влияет наличие частых командировок на выбытие из штата сотрудников исследуемой группы? 

In [12]:
df_research = df[(df['Age'] >= 30) & (df['Age'] <= 43) & (df['Gender'] == 'Male')]
df_research_attr = df_research[df_research['Attrition'] == 'Yes']
df_research_attr_tr = df_research_attr[df_research_attr['BusinessTravel'] == 'Travel_Rarely']
df_research_attr_tf = df_research_attr[df_research_attr['BusinessTravel'] == 'Travel_Frequently']
df_research_attr_nt = df_research_attr[df_research_attr['BusinessTravel'] == 'Non-Travel']

len(df_research_attr_tr) * 100 / len(df_research_attr), len(df_research_attr_tf) * 100 / len(df_research_attr), len(df_research_attr_nt) * 100 / len(df_research_attr)


(66.66666666666667, 25.757575757575758, 7.575757575757576)

ответ: в исследуемой группе наиболее часто выбывали сотрудники с редкими командировками

#### 4.	Узнай доли выбывших сотрудников исследуемой группы по каждому отделу

In [25]:
attr_dep = df_research_attr['Department'].value_counts() / df_research['Department'].value_counts()
attr_dep

Department
Research & Development    0.126246
Sales                     0.172185
Human Resources           0.080000
Name: count, dtype: float64

ответ: больше всего в исследуемой группе выбыло сотрудников из отдела "Sales", меньше всего из отдела "Human Resources"

#### 5.	Узнай наиболее часто встречаемую должность среди тех, кто выбыл из штата  в исследуемой группе.

In [14]:
df_research_attr['JobRole'].value_counts()

JobRole
Sales Executive              22
Laboratory Technician        17
Research Scientist           12
Manufacturing Director        5
Sales Representative          4
Healthcare Representative     3
Human Resources               2
Manager                       1
Name: count, dtype: int64

ответ: в исследуемой группе чаще всего выбывают сотрудники должности "Sales Executive"

#### 6.	Сотрудники с каким семейным положением в исследуемой группе выбывают из штата чаще?

In [15]:
df_research_attr.groupby(['MaritalStatus']).size()

MaritalStatus
Divorced    11
Married     28
Single      27
dtype: int64

ответ: видно, что женатых и одиноких мужчин среди выбывших в исследуемой группе почти поровну.

#### 7.	Какой средний ежемесячный доход в исследуемой группе? Узнай долю сотрудников, имеющих сверхурочные работы, среди тех, кто выбыл из штата в исследуемой группе  и чей ежемесячный доход ниже найденного среднего.

In [16]:
mean_mr = df_research['MonthlyIncome'].mean().round(2)
mean_mr

6164.61

In [17]:
df_research['MonthlyIncome'].describe()

count      477.000000
mean      6164.607966
std       4082.117685
min       1081.000000
25%       3057.000000
50%       5126.000000
75%       7596.000000
max      19833.000000
Name: MonthlyIncome, dtype: float64

In [18]:
df_research_mr = df_research_attr[(df_research_attr['MonthlyIncome'] <= mean_mr)]
df_research_mrover = df_research_mr[(df_research_mr['OverTime'] == 'Yes')]
len(df_research_mrover) / len(df_research_mr)

0.5652173913043478

ответ: средний ежемесячный доход в исследуемой группе равен 6164.61. Доля сотрудников, имеющих сверхурочные работы среди тех, кто выбыл из штата в исследуемой группе  и ежемесячный доход ниже найденного среднего равна 0,5652.

#### 8.	Сколько сотрудников исследуемой группы служат в текущей компании меньше 10 лет?  Какая доля таких сотрудников выбыла из штата?

In [28]:
df_research_10 = df_research[(df_research['YearsAtCompany'] <= 10)]
len(df_research_10)

390

In [29]:
len(df_research_attr[(df_research_attr['YearsAtCompany'] <= 10)]) / len(df_research_10)

0.15128205128205127

ответ: меньше 10 лет в исследуемой группе работают 390 человек. Доля выбывших сотрудников среди них равна 0,15

#### 9. Правда ли, что среди выбывших сотрудников исследуемой группы больше всего тех, чье образование ниже магистра?

In [21]:
df_research_attr.groupby(['Education']).size()

Education
1     5
2     9
3    28
4    22
5     2
dtype: int64

ответ: да, среди выбывших сотрудников исследуемой группы больше людей, чье образование ниже магистра (1, 2, 3)

#### 10.	 Среди выбывших сотрудников исследуемой группы с какой областью образования больше? 

In [22]:
df_research_attr['EducationField'].value_counts()

EducationField
Life Sciences       23
Medical             22
Marketing            8
Other                6
Technical Degree     5
Human Resources      2
Name: count, dtype: int64

ответ: среди выбывших сотрудников исследуемой группы больше сотрудников с областью образования "Естественные науки"

#### 11. Правда ли, что среди выбывших сотрудников исследуемой группы больше всего тех, у кого низкая удовлетворенность работой? 

In [23]:
df_research_attr['JobSatisfaction'].value_counts()


JobSatisfaction
1    22
3    19
4    16
2     9
Name: count, dtype: int64

ответ: да